uefi/src/proto/shell/mod.rs

```diff
@@ -54,4 +95,159 @@ impl Shell {
         let dir_ptr: *const Char16 = directory.map_or(ptr::null(), |x| x.as_ptr());
         unsafe { (self.0.set_cur_dir)(fs_ptr.cast(), dir_ptr.cast()) }.to_result()
     }
+
+    /// Gets the value of the specified environment variable
+    ///
+    /// # Arguments
+    ///
+    /// * `name` - The environment variable name of which to retrieve the
+    ///   value.
+    ///
+    /// # Returns
+    ///
+    /// * `Some(<env_value>)` - &CStr16 containing the value of the
+    ///   environment variable
+    /// * `None` - If environment variable does not exist
+    #[must_use]
+    pub fn var(&self, name: &CStr16) -> Option<&CStr16> {
+        let name_ptr: *const Char16 = name.as_ptr();
+        let var_val = unsafe { (self.0.get_env)(name_ptr.cast()) };
+        if var_val.is_null() {
+            None
+        } else {
+            unsafe { Some(CStr16::from_ptr(var_val.cast())) }
+        }
+    }
+
+    /// Gets an iterator over the names of all environment variables
+    ///
+    /// # Returns
+    ///
+    /// * `Vars` - Iterator over the names of the environment variables
+    #[must_use]
+    pub fn vars(&self) -> Vars<'_, Self> {
+        let env_ptr = unsafe { (self.0.get_env)(ptr::null()) };
+        Vars {
+            names: env_ptr.cast::<Char16>(),
+            protocol: self,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Sets the environment variable
+    ///
+    /// # Arguments
+    ///
+    /// * `name` - The environment variable for which to set the value
+    /// * `value` - The new value of the environment variable
+    /// * `volatile` - Indicates whether the variable is volatile or
+    ///   not
+    ///
+    /// # Returns
+    ///
+    /// * `Status::SUCCESS` - The variable was successfully set
+    pub fn set_var(&self, name: &CStr16, value: &CStr16, volatile: bool) -> Result {
+        let name_ptr: *const Char16 = name.as_ptr();
+        let value_ptr: *const Char16 = value.as_ptr();
+        unsafe { (self.0.set_env)(name_ptr.cast(), value_ptr.cast(), volatile) }.to_result()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use alloc::collections::BTreeMap;
+    use alloc::vec::Vec;
+    use uefi::cstr16;
+
+    struct ShellMock<'a> {
+        inner: BTreeMap<&'a CStr16, &'a CStr16>,
+    }
+
+    impl<'a> ShellMock<'a> {
+        fn new(names: Vec<&'a CStr16>, values: Vec<&'a CStr16>) -> ShellMock<'a> {
```

Comment:
At least

```suggestion
        fn new(names: &[&'a CStr16], values: &[&'a CStr16]) -> ShellMock<'a> {
```

you should also assert that both's len() equals.

But **much better** would be:


```suggestion
        fn new(pairs: impl IntoIterator<Item = (&CStr16, &CStr16)>) -> ShellMock<'a> {
```

as this is your unit test where you have control over. Your initial strategy is way too complicated; it can be done much simpler :)